### PR TITLE
Abort configure if static component depends on dso common

### PR DESCRIPTION
**Pull request by bwbarrett**
Another take on fixing https://github.com/open-mpi/ompi/issues/9451 (original attempt https://github.com/open-mpi/ompi/pull/10710).  This avoids running the evaluation logic for build type multiple times, and hides all the logic in a central place.

### CODE REVIEW

File: config/opal_mca.m4 [hunk 1]
```diff
@@ -720,27 +759,31 @@ AC_DEFUN([MCA_COMPONENT_COMPILE_MODE],[
...
     AC_MSG_CHECKING([for MCA component $2:$3 compile mode])
     if test "$DIRECT_$2" = "$3" ; then
-        AC_MSG_RESULT([$$4 - direct])
+        AC_MSG_RESULT([$compile_mode_internal_tmp - direct])
     else
-        AC_MSG_RESULT([$$4])
+        AC_MSG_RESULT([$compile_mode_internal_tmp])
     fi
+
+    OPAL_VAR_SCOPE_POP([compile_mode_internal_tmp])
```
Comment (jsquyres) [inline config/opal_mca.m4:786]: `OPAL_VAR_SCOPE_POP` doesn't take any arguments -- it unwinds *all* the variable names from the most recent `OPAL_VAR_SCOPE_PUSH`.

File: config/opal_mca.m4 [hunk 2]
```diff
@@ -706,10 +703,52 @@ AC_DEFUN([MCA_CONFIGURE_ALL_CONFIG_COMPONENTS],[
...
+	[AS_LITERAL_IF([$3],
+             [m4_ifdef([MCA_$1_$2_$3_COMPILE_MODE],
+                  [dnl We introduced caching of this check after setting the compile
+                   dnl mode by the substitute macro was common, and there was not a
+                   dnl polymorphic variable assumption in all those macros, so we use
+                   dnl a temporary with a fixed name.
+                   OPAL_VAR_SCOPE_PUSH([component_compile_mode_tmp])
+                   MCA_$1_$2_$3_COMPILE_MODE([$1], [$2], [$3], [component_compile_mode_tmp])
+                   AS_VAR_COPY([compile_mode_cv], [$component_compile_mode_tmp])
+                   OPAL_VAR_SCOPE_POP([component_compile_mode_tmp])],
```
Comment (jsquyres) [inline config/opal_mca.m4:729]: `OPAL_VAR_SCOPE_POP` does not take any arguments.

File: config/opal_mca.m4 [hunk 3]
```diff
@@ -1033,3 +1033,29 @@ AC_DEFUN([OPAL_MCA_MAKE_DIR_LIST],[
...
+    MCA_COMPONENT_COMPILE_MODE([$4], [$5], [$6], [dependency_compile_mode])
+
+    AS_IF([test "${component_compile_mode}" = "static" -a "${dependency_compile_mode}" = "dso"],
+         [AC_MSG_ERROR([Component $2:$3 is set to build as a
+static component, but its dependency $5:$6 is set to build as
+a dynamic component.  This configuration is not supported.  Please
+either build $5:$6 as a static component or build $2:$3 as a dynamic
+component.])])
+
+    OPAL_VAR_SCOPE_POP([component_compile_mode dependency_compile_mode])
```
Comment (jsquyres) [inline config/opal_mca.m4:1060]: `OPAL_VAR_SCOPE_POP` does not take any arguments.

**Comment by bwbarrett** [comment]
Rebased and addressed Jeff's `OPAL_VAR_SCOPE_POP` comments.

**Review by jsquyres — APPROVED** [review]

**Pull request merged**